File: Graph/__1000__/1100/1100.cpp
#include <iostream>
#include <vector>
#include <utility>
#include <tuple>
#include <string>
#include <unordered_set>
#include <queue>
#include <sstream>

using namespace std;

// Definindo um tipo para coordenadas
using TCoord = pair<int, int>;

// Hash para usar pair<int,int> em unordered_set
struct PairHash {
    size_t operator()(const TCoord& coord) const {
        return hash<int>()(coord.first) ^ (hash<int>()(coord.second) << 1);
    }
};

typedef unordered_set<TCoord, PairHash> HashSet;
typedef pair<int, TCoord> Moves;  // número de movimentos + coordenada

// variações de coordenadas para cada salto do cavalo
const vector<pair<int, int>> DELTA = {
    { 2,  1}, { 1,  2}, {-1,  2}, {-2,  1},
    {-2, -1}, {-1, -2}, { 1, -2}, { 2, -1}
};

// checa se as coordenadas x, y estão dentro do limite do board
bool inBound(int x, int y) {
    return 0 <= x && x < 8 && 0 <= y && y < 8;
}

// Transforma a posição (ex: "a1") em coordenadas (linha, coluna)
TCoord GetCoord(const string& position) {
    int col = position[0] - 'a'; // 'a' → 0, 'b' → 1, ...
    int row = position[1] - '1'; // '1' → 0, '2' → 1, ...
    return {row, col};
}

// Breadth_First_Search - BFS 
int BFS(const TCoord& source, const TCoord& target) {
    
    HashSet visiteds;                      // set das coordenadas visitadas
    queue<Moves> q;                        // fila de jogadas

    q.push({0, source});                   // adiciona primeira jogada (0 movimentos, coord inicial)
    visiteds.insert(source);               // marca coord inicial como visitada

    while (!q.empty()) {
        Moves move = q.front();
        q.pop();

        int moves = move.first;
        int x = move.second.first;
        int y = move.second.second;

        if (move.second == target) return moves; // chegou no alvo

        for (auto [dx, dy] : DELTA) {            // testa cada uma das jogadas possíveis
            int nx = x + dx;
            int ny = y + dy;

            TCoord next = {nx, ny};

            if (inBound(nx, ny) && visiteds.find(next) == visiteds.end()) {
                visiteds.insert(next);
                q.push({moves + 1, next});
            }
        }
    }

    return -1; // só por garantia.
}

int main() {
    string source, target;
    while (cin >> source >> target) {
        TCoord source_coord = GetCoord(source);
        TCoord target_coord = GetCoord(target);

        int moves = BFS(source_coord, target_coord);

        cout << "To get from " << source << " to " << target
             << " takes " << moves << " knight moves." << endl;
    }
    return 0;
}

File: Graph/__1000__/1100/1100.cs
using System;
using System.Collections.Generic;

public class URI
{
    // Lista de deltas do cavalo (dx, dy)
    private static readonly (int dx, int dy)[] deltas =
    {
        ( 2,  1), ( 1,  2), (-1,  2), (-2,  1),
        (-2, -1), (-1, -2), ( 1, -2), ( 2, -1)
    };

    public static void Main(string[] args)
    {
        string line;
        while ((line = Console.ReadLine()) != null && line != "")
        {
            var input = line.Split(' ');
            var source = GetCoord(input[0]);
            var target = GetCoord(input[1]);

            int moves = BFS(source, target);

            Console.WriteLine(
                $"To get from {input[0]} to {input[1]} takes {moves} knight moves."
            );
        }
    }

    // Transforma a posição (ex: "a1") em coordenadas (linha, coluna)
    public static (int x, int y) GetCoord(string position)
    {
        int col = position[0] - 'a'; // a→0, b→1, ...
        int row = position[1] - '1'; // 1→0, 2→1, ...
        return (row, col);
    }

    // BFS para calcular menor número de movimentos do cavalo
    public static int BFS((int x, int y) source, (int x, int y) target)
    {
        var queue = new Queue<(int moves, int x, int y)>();
        var visited = new HashSet<(int x, int y)>();

        queue.Enqueue((0, source.x, source.y));
        visited.Add(source);

        while (queue.Count > 0)
        {
            var (moves, x, y) = queue.Dequeue();

            if (x == target.x && y == target.y)
                return moves;

            for (int i = 0; i < deltas.Length; i++)
            {
                int nx = x + deltas[i].dx;
                int ny = y + deltas[i].dy;

                if (InBound(nx, ny) && !visited.Contains((nx, ny)))
                {
                    visited.Add((nx, ny));
                    queue.Enqueue((moves + 1, nx, ny));
                }
            }
        }

        return -1; // Nunca deve acontecer no tabuleiro 8x8
    }

    // Verifica se está dentro dos limites do tabuleiro 8x8
    public static bool InBound(int x, int y)
    {
        return (0 <= x && x < 8 && 0 <= y && y < 8);
    }
}